infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/OpenShiftMachine.java

```diff
@@ -272,6 +217,62 @@ public void onClose(KubernetesClientException cause) {
         return encoded;
     }
 
+    private List<Service> getMatchedServices() {
+        Pod pod = getPod();
+        Container container = pod.getSpec()
+                                 .getContainers()
+                                 .stream()
+                                 .filter(c -> containerName.equals(c.getName()))
+                                 .findAny()
+                                 .orElseThrow(
+                                         () -> new IllegalStateException("Corresponding pod for OpenShift machine doesn't exist."));
+
+        try (OpenShiftClient client = clientFactory.create()) {
+            return client.services()
+                         .inNamespace(projectName)
+                         .list()
+                         .getItems()
+                         .stream()
+                         .filter(service -> isExposedByService(pod, service))
+                         .filter(service -> isExposedByService(container, service))
+                         .collect(Collectors.toList());
+        }
+    }
+
+    private static boolean isExposedByService(Pod pod, Service service) {
+        Map<String, String> labels = pod.getMetadata().getLabels();
+        Map<String, String> selectorLabels = service.getSpec().getSelector();
+        if (labels == null) {
+            return false;
+        }
+        for (Map.Entry<String, String> selectorLabelEntry : selectorLabels.entrySet()) {
+            if (!selectorLabelEntry.getValue().equals(labels.get(selectorLabelEntry.getKey()))) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private static boolean isExposedByService(Container container, Service service) {
+        for (ServicePort servicePort : service.getSpec().getPorts()) {
+            IntOrString targetPort = servicePort.getTargetPort();
+            if (targetPort.getIntVal() != null) {
```

Comment:
Is it possible to get targetPort without int val part?
I thought that it's either in format _8000_ or _8000/tcp_